alphafold3_pytorch/common/dna_constants.py

```diff
@@ -96,4 +96,135 @@
 resnames = [restype_1to3[r] for r in restypes] + [unk_restype]
 
 # This represents the residue chemical type (i.e., `chemtype`) index of DNA residues.
-chemtype_num = rna_constants.chemtype_num + 1  # := 2.
\ No newline at end of file
+chemtype_num = rna_constants.chemtype_num + 1  # := 2.
+
+# A compact atom encoding with 23 columns for DNA residues.
+# pylint: disable=line-too-long
+# pylint: disable=bad-whitespace
+restype_name_to_compact_atom_names = {
+    "DA": [
+        "C3'",
+        "C4'",
+        "O4'",
+        "C2'",
+        "C1'",
+        "C5'",
+        "O3'",
+        "O5'",
+        "P",
+        "OP1",
+        "OP2",
+        "N9",
+        "C4",
+        "N1",
+        "N3",
+        "N6",
+        "N7",
+        "C2",
+        "C5",
+        "C6",
+        "C8",
+        "",
+        "",
+    ],
+    "DC": [
+        "C3'",
+        "C4'",
```

Comment:
@amorehead so just to reiterate conversation on discord, would be helpful to have the bonds for these atoms specified somewhere.

Reply:
Right, I was just thinking about this. I'll need to get back to you on this, as these compact atom positions were simply pulled from structure prediction codebases that didn't mess with bond graphs for each residue. Actually, on my local machine, I was analyzing the differences between the atoms in e.g., `DT` (in `dna_constants.py`) and the atoms in `T` (in `life.py`), and I noticed that `dna_constants.py` has one fewer carbon atom in `DT` compared to `life.py`'s SMILES string for `T` (i.e., 10 in `life.py`, 9 in `dna_constants.py`).

Reply:
Actually, on my local machine, I was analyzing the differences between the atoms in e.g., `DT` (in `dna_constants.py`) and the atoms in `T` (in `life.py`), and I noticed that `dna_constants.py` has one fewer carbon atom in `DT` compared to `life.py`'s SMILES string for `T` (i.e., 10 in `life.py`, 9 in `dna_constants.py`).

However, this is the only difference (in terms of atom compositions) I could find between `life.py` and `{protein,rna,dna,ligand}_constants.py`.

Reply:
Concretely, here's what the atom composition breakdowns look like when comparing `life.py` and the constants scripts:
[life_dna_residue_atoms.json](https://github.com/user-attachments/files/16203424/life_dna_residue_atoms.json)
[life_protein_residue_atoms.json](https://github.com/user-attachments/files/16203425/life_protein_residue_atoms.json)
[life_rna_residue_atoms.json](https://github.com/user-attachments/files/16203426/life_rna_residue_atoms.json)
[rc_dna_residue_atoms.json](https://github.com/user-attachments/files/16203427/rc_dna_residue_atoms.json)
[rc_protein_residue_atoms.json](https://github.com/user-attachments/files/16203428/rc_protein_residue_atoms.json)
[rc_rna_residue_atoms.json](https://github.com/user-attachments/files/16203429/rc_rna_residue_atoms.json)

Reply:
> Right, I was just thinking about this. I'll need to get back to you on this, as these compact atom positions were simply pulled from structure prediction codebases that didn't mess with bond graphs for each residue.

oh nice, so the bond graphs are defined somewhere? that's really all I need

Reply:
On second thought (since I'm not sure where to find these bond graphs), I think it might instead make more sense to look at the CCD's SMILES string for each amino acid/nucleotide. If there are any SMILES strings we can consider "official", it should be these strings, since the CCD is curated by the Protein Data Bank. Plus, since each amino acid/nucleotide has an associated SMILES string in the PDB CCD, there should also be a canonical "structure" of each of these residues that we can reference here. I'll take a look at the PDB for these structure now.

Reply:
Ah yes, it looks like the CIF definition files for each of these amino acids/nucleotides contain the canonical "structure" of the residue, and the same page contains the isomeric SMILES for each residue as well. Now the only question remaining is, how does one map the canonical structure (i.e., atoms) of each residue one-to-one to the SMILES string, since a SMILES string only contains primitive atom types (e.g., N and C atoms)?

Reply:
Oh, perhaps the following fields in these CIF definition files can be used to infer this atom mapping? For example, these for ALA are:

```
# 
loop_
_chem_comp_atom.comp_id 
_chem_comp_atom.atom_id 
_chem_comp_atom.alt_atom_id 
_chem_comp_atom.type_symbol 
_chem_comp_atom.charge 
_chem_comp_atom.pdbx_align 
_chem_comp_atom.pdbx_aromatic_flag 
_chem_comp_atom.pdbx_leaving_atom_flag 
_chem_comp_atom.pdbx_stereo_config 
_chem_comp_atom.pdbx_backbone_atom_flag 
_chem_comp_atom.pdbx_n_terminal_atom_flag 
_chem_comp_atom.pdbx_c_terminal_atom_flag 
_chem_comp_atom.model_Cartn_x 
_chem_comp_atom.model_Cartn_y 
_chem_comp_atom.model_Cartn_z 
_chem_comp_atom.pdbx_model_Cartn_x_ideal 
_chem_comp_atom.pdbx_model_Cartn_y_ideal 
_chem_comp_atom.pdbx_model_Cartn_z_ideal 
_chem_comp_atom.pdbx_component_atom_id 
_chem_comp_atom.pdbx_component_comp_id 
_chem_comp_atom.pdbx_ordinal 
ALA N   N   N 0 1 N N N Y Y N 2.281  26.213 12.804 -0.966 0.493  1.500  N   ALA 1  
ALA CA  CA  C 0 1 N N S Y N N 1.169  26.942 13.411 0.257  0.418  0.692  CA  ALA 2  
ALA C   C   C 0 1 N N N Y N Y 1.539  28.344 13.874 -0.094 0.017  -0.716 C   ALA 3  
ALA O   O   O 0 1 N N N Y N Y 2.709  28.647 14.114 -1.056 -0.682 -0.923 O   ALA 4  
ALA CB  CB  C 0 1 N N N N N N 0.601  26.143 14.574 1.204  -0.620 1.296  CB  ALA 5  
ALA OXT OXT O 0 1 N Y N Y N Y 0.523  29.194 13.997 0.661  0.439  -1.742 OXT ALA 6  
ALA H   H   H 0 1 N N N Y Y N 2.033  25.273 12.493 -1.383 -0.425 1.482  H   ALA 7  
ALA H2  HN2 H 0 1 N Y N Y Y N 3.080  26.184 13.436 -0.676 0.661  2.452  H2  ALA 8  
ALA HA  HA  H 0 1 N N N Y N N 0.399  27.067 12.613 0.746  1.392  0.682  HA  ALA 9  
ALA HB1 1HB H 0 1 N N N N N N -0.247 26.699 15.037 1.459  -0.330 2.316  HB1 ALA 10 
ALA HB2 2HB H 0 1 N N N N N N 0.308  25.110 14.270 0.715  -1.594 1.307  HB2 ALA 11 
ALA HB3 3HB H 0 1 N N N N N N 1.384  25.876 15.321 2.113  -0.676 0.697  HB3 ALA 12 
ALA HXT HXT H 0 1 N Y N Y N Y 0.753  30.069 14.286 0.435  0.182  -2.647 HXT ALA 13 
# 
loop_
_chem_comp_bond.comp_id 
_chem_comp_bond.atom_id_1 
_chem_comp_bond.atom_id_2 
_chem_comp_bond.value_order 
_chem_comp_bond.pdbx_aromatic_flag 
_chem_comp_bond.pdbx_stereo_config 
_chem_comp_bond.pdbx_ordinal 
ALA N   CA  SING N N 1  
ALA N   H   SING N N 2  
ALA N   H2  SING N N 3  
ALA CA  C   SING N N 4  
ALA CA  CB  SING N N 5  
ALA CA  HA  SING N N 6  
ALA C   O   DOUB N N 7  
ALA C   OXT SING N N 8  
ALA CB  HB1 SING N N 9  
ALA CB  HB2 SING N N 10 
ALA CB  HB3 SING N N 11 
ALA OXT HXT SING N N 12 
```

Reply:
yes that's it! thank you

xyz is enough to derive distance and infer the bonds

Reply:
also, feel free to reach out to me on Signal if you feel like it would be faster to communicate that way

Reply:
Np. I'm still looking into these PDB ligands for each amino acid/nucleotide residue to make sure the canonical structures of each residue match what we have in the constants.py files.